File: src/usage/meters.ts
// src/usage/meters.ts
//
// LIFTED VERBATIM from cloud-portal
//   branch:  feat/patch-dynamic-composition
//   path:    app/modules/usage/meters.ts
// (branch version — includes the agent-framework `tool-invocations` meter)
// -----------------------------------------------------------------------
//
// Canonical meter and resource identifiers for the AI assistant. These
// MUST stay in sync with the assistant ServiceConfiguration in
// config/services/services_v1alpha1_serviceconfiguration_assistant.yaml,
// which the services-operator fans out into
// billing.miloapis.com/MeterDefinition and MonitoredResourceType
// objects in Milo.
//
// metrics[].name, .kind, .unit and monitoredResourceTypes[].type,
// .gvk are immutable on the ServiceConfiguration side once Published;
// treat the names below as wire constants and version any breaking
// change as a new metric name (e.g. ".../v2").

export const ASSISTANT_SERVICE_NAME = 'assistant.miloapis.com';

/** Group/Kind that emits assistant usage events. */
export const ASSISTANT_RESOURCE_GROUP = 'assistant.miloapis.com';
export const ASSISTANT_RESOURCE_KIND = 'Conversation';

/** Canonical meter names. Reverse-DNS path under the service name. */
export const ASSISTANT_METERS = {
  inputTokens: 'assistant.miloapis.com/conversation/input-tokens',
  outputTokens: 'assistant.miloapis.com/conversation/output-tokens',
  cacheReadTokens: 'assistant.miloapis.com/conversation/cache-read-tokens',
  cacheWriteTokens: 'assistant.miloapis.com/conversation/cache-write-tokens',
  messages: 'assistant.miloapis.com/conversation/messages',
  // Agent-framework provider tools (AGENT_FRAMEWORK_ENABLED): one Delta
  // event per provider-tool invocation, dimensioned by the provider's
  // reverse-DNS `service` name (AgentBinding spec.serviceName).
  toolInvocations: 'assistant.miloapis.com/conversation/tool-invocations',
} as const;

export type AssistantMeterKey = keyof typeof ASSISTANT_METERS;

File: internal/usage/meters.go
package usage

// Canonical meter and resource identifiers for the AI assistant, ported
// verbatim from the TS service's src/usage/meters.ts.
//
// These MUST stay in sync with the assistant ServiceConfiguration that the
// services-operator fans out into billing.miloapis.com/MeterDefinition and
// MonitoredResourceType objects in Milo. metric names, kind, unit and the
// monitored-resource type/gvk are immutable once Published — treat the values
// below as wire constants and version any breaking change as a new metric name.

// ServiceName is the reverse-DNS service identifier for the assistant.
const ServiceName = "assistant.miloapis.com"

// Group/Kind that emits assistant usage events.
const (
	ResourceGroup = "assistant.miloapis.com"
	ResourceKind  = "Conversation"
)

// Canonical meter names (reverse-DNS paths under the service name).
const (
	MeterInputTokens      = "assistant.miloapis.com/conversation/input-tokens"
	MeterOutputTokens     = "assistant.miloapis.com/conversation/output-tokens"
	MeterCacheReadTokens  = "assistant.miloapis.com/conversation/cache-read-tokens"
	MeterCacheWriteTokens = "assistant.miloapis.com/conversation/cache-write-tokens"
	MeterMessages         = "assistant.miloapis.com/conversation/messages"
	// MeterToolInvocations is a Delta counter: one event per provider-tool
	// invocation, dimensioned by the provider's reverse-DNS service name.
	MeterToolInvocations = "assistant.miloapis.com/conversation/tool-invocations"
)
